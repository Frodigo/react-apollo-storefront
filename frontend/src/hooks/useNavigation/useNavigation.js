import { useMemo } from 'react';
import { useQuery } from '@apollo/client';
import { GET_NAVIGATION } from '../../queries/category'

/**
 * The useNavigation hook provides data and business logic for the Navigation component
 *
 *
 * @return {
 *  navigationData {array} - array of categories to display on a storefront
 * }
 */
export const useNavigation = () => {
    /**
     * Get category data from API
     */
    const { data } = useQuery(GET_NAVIGATION, {
        fetchPolicy: 'cache-and-network',
        nextFetchPolicy: 'cache-first'
    });

    /**
     * Check if category should be visible on the storefront.
     *
     * @param {NavigationItem} category
     * @returns {boolean}
     */
    const shouldRenderNavigationItem = category => {
        return !!category.include_in_menu;
    };


    /**
     * Preparing navigation data
     * Sorts by category.position
     * Filtering by category.include_in_menu flag
     * Adds categoryUrl field
     *
     * @returns {array}
     */
    const navigationData = useMemo(() => {
        const rootCategory = data && data.categoryList ? data.categoryList[0] : null;

        if (rootCategory) {
            return rootCategory.children
                .filter(category => shouldRenderNavigationItem(category))
                .sort((a, b) => (a.position > b.position ? 1 : -1))
                .map(category => {
                    const categoryUrl = `/category/${category.url_path}${category.url_suffix}`

                    return {
                        ...category,
                        categoryUrl
                    }
                })
        }

        return [];
    }, [data]);

    return {
        navigationData
    }
}

export default useNavigation;
