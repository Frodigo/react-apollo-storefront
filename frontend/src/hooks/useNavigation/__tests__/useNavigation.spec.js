import React from 'react';
import { render, FindBy } from '@testing-library/react';
import { useQuery } from '@apollo/client';
import useNavigation from '../useNavigation'

jest.mock('@apollo/client');

const setupHook = () => {
    const returnVal = {}

    const TestComponent = () => {
        Object.assign(returnVal, useNavigation());

        return null
    }

    render(<TestComponent />)

    return returnVal
}


test('Should render proper shape', () => {
    useQuery.mockReturnValue(
        {
            "data": {
                "categoryList": [
                    {
                        "id": 2,
                        "name": "Default Category",
                        "children": [
                            {
                                "id": 38,
                                "include_in_menu": 1,
                                "name": "What's New",
                                "position": 10,
                                "url_path": "what-is-new",
                                "url_suffix": ".html"
                            },
                            {
                                "id": 20,
                                "include_in_menu": 0,
                                "name": "Women",
                                "position": 2,
                                "url_path": "women",
                                "url_suffix": ".html"
                            },
                            {
                                "id": 11,
                                "include_in_menu": 1,
                                "name": "Men",
                                "position": 3,
                                "url_path": "men",
                                "url_suffix": ".html"
                            },
                            {
                                "id": 3,
                                "include_in_menu": 1,
                                "name": "Gear",
                                "position": 4,
                                "url_path": "gear",
                                "url_suffix": ".html"
                            },
                            {
                                "id": 9,
                                "include_in_menu": 1,
                                "name": "Training",
                                "position": 5,
                                "url_path": "training",
                                "url_suffix": ".html"
                            },
                            {
                                "id": 37,
                                "include_in_menu": 1,
                                "name": "Sale",
                                "position": 1,
                                "url_path": "sale",
                                "url_suffix": ".html"
                            }
                        ]
                    }
                ]
            }
        }
    );

    const hookProps = setupHook();

    expect(hookProps).toMatchSnapshot();
});


test('Should sort items', () => {
    useQuery.mockReturnValue(
        {
            "data": {
                "categoryList": [
                    {
                        "id": 2,
                        "name": "Default Category",
                        "children": [
                            {
                                "id": 38,
                                "include_in_menu": 1,
                                "name": "What's New",
                                "position": 10,
                                "url_path": "what-is-new",
                                "url_suffix": ".html"
                            },
                            {
                                "id": 20,
                                "include_in_menu": 1,
                                "name": "Women",
                                "position": 1,
                                "url_path": "women",
                                "url_suffix": ".html"
                            }
                        ]
                    }
                ]
            }
        }
    );

    const hookProps = setupHook();

    expect(hookProps.navigationData[0].name).toEqual('Women');
});

test('Should not render items that are not included in menu', () => {
    useQuery.mockReturnValue(
        {
            "data": {
                "categoryList": [
                    {
                        "id": 2,
                        "name": "Default Category",
                        "children": [
                            {
                                "id": 38,
                                "include_in_menu": 1,
                                "name": "What's New",
                                "position": 10,
                                "url_path": "what-is-new",
                                "url_suffix": ".html"
                            },
                            {
                                "id": 20,
                                "include_in_menu": 0,
                                "name": "Women",
                                "position": 1,
                                "url_path": "women",
                                "url_suffix": ".html"
                            }
                        ]
                    }
                ]
            }
        }
    );

    const hookProps = setupHook();

    expect(hookProps.navigationData.length).toEqual(1);

});
