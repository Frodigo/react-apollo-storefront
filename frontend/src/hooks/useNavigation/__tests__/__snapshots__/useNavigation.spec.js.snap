// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render proper shape 1`] = `
Object {
  "navigationData": Array [
    Object {
      "categoryUrl": "/category/sale.html",
      "id": 37,
      "include_in_menu": 1,
      "name": "Sale",
      "position": 1,
      "url_path": "sale",
      "url_suffix": ".html",
    },
    Object {
      "categoryUrl": "/category/men.html",
      "id": 11,
      "include_in_menu": 1,
      "name": "Men",
      "position": 3,
      "url_path": "men",
      "url_suffix": ".html",
    },
    Object {
      "categoryUrl": "/category/gear.html",
      "id": 3,
      "include_in_menu": 1,
      "name": "Gear",
      "position": 4,
      "url_path": "gear",
      "url_suffix": ".html",
    },
    Object {
      "categoryUrl": "/category/training.html",
      "id": 9,
      "include_in_menu": 1,
      "name": "Training",
      "position": 5,
      "url_path": "training",
      "url_suffix": ".html",
    },
    Object {
      "categoryUrl": "/category/what-is-new.html",
      "id": 38,
      "include_in_menu": 1,
      "name": "What's New",
      "position": 10,
      "url_path": "what-is-new",
      "url_suffix": ".html",
    },
  ],
}
`;
